src/main/java/com/google/edith/services/StoreReceiptService.java

```diff
@@ -0,0 +1,159 @@
+import com.google.gson.JsonParser;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.Optional;
+import javax.servlet.http.HttpServletRequest;
+
+public final class StoreReceiptService implements StoreReceiptInterface {
+
+  private final DatastoreService datastore;
+
+  public StoreReceiptService(DatastoreService datastore) {
+    this.datastore = datastore;
+  }
+
+  @Override
+  public void storeEntites(Receipt receipt) {
+    storeReceiptEntity(receipt);
+  }
+
+  @Override
+  public Receipt parseReceiptFromForm(HttpServletRequest request) throws IOException {
+    BufferedReader bufferedReader = request.getReader();
+    Gson gson = new GsonBuilder().registerTypeAdapter(Item.class, new ItemDeserializer()).create();
+    JsonParser parser = new JsonParser();
+    JsonObject json = (JsonObject) parser.parse(bufferedReader);
+    return gson.fromJson(json, Receipt.class);
+  }
+
+  /**
+   * Receives Receipt object and creates entity of type Receipt and stores it in Datastore.
+   *
+   * @param receipt - object which holds info of parsed file.
+   */
+  private void storeReceiptEntity(Receipt receipt) {
+    String userId = receipt.getUserId();
+    String storeName = receipt.getStoreName();
+    String date = receipt.getDate();
+    String name = receipt.getName();
+    String fileUrl = receipt.getFileUrl();
+    float totalPrice = receipt.getTotalPrice();
+
+    Optional<Entity> optEntity = getUserInfoEntity(userId);
+    Entity userInfoEntity = optEntity.get();
+    Entity receiptEntity = new Entity("Receipt", userInfoEntity.getKey());
+    receiptEntity.setProperty("userId", userId);
+    receiptEntity.setProperty("storeName", storeName);
+    receiptEntity.setProperty("date", date);
+    receiptEntity.setProperty("name", name);
+    receiptEntity.setProperty("fileUrl", fileUrl);
+    receiptEntity.setProperty("totalPrice", totalPrice);
+    datastore.put(receiptEntity);
+    storeReceiptItemsEntity(receipt, receiptEntity);
+  }
+
+  /**
+   * Stores parsed item from the form with receiptEntity as a parent in the datastore.
+   *
+   * @param items - request which contains the form body.
+   * @param receiptEntity - request which contains the form body.
+   */
+  private void storeReceiptItemsEntity(Receipt receipt, Entity receiptEntity) {
+    Item[] items = receipt.getItems();
+    for (Item item : items) {
+      String userId = item.userId();
+      String itemName = item.name();
+      double price = item.price();
+      long quantity = item.quantity();
+      String category = item.category();
+      String expireDate = item.expiration();
+
+      Entity itemEntity = new Entity("Item", receiptEntity.getKey());
+      itemEntity.setProperty("userId", userId);
+      itemEntity.setProperty("name", itemName);
+      itemEntity.setProperty("quantity", quantity);
+      itemEntity.setProperty("category", category);
+      itemEntity.setProperty("expireDate", expireDate);
```

Comment:
This can be extracted to a separate method:

```java
private static Entity createItemEntity(Item item) {
   ...
}
```

then it would be possible to map these

```java
receipt.getItems().map(StoreReceiptServiceImpl::createItemEntity).forEach(datastore::put);
```

Reply:
Thanks for the edits.